### PR TITLE
Setup github action for linting C++ codebase

### PR DESCRIPTION
**Pull request by hungpham2511**
Related to #84 

In this PR a linting CI job is added via github action using clang-tidy. Current the following checks are made:
```
        WARN_FLAGS=-checks="clang-analyzer-*,clang-analyzer-cplusplus*,cppcoreguidelines-*,performance-*,modernize-*,readability-*"
        ERR_FLAGS=-warnings-as-errors="clang-analyzer-*,clang-analyzer-cplusplus*,cppcoreguidelines-*,modernize-*"
        clang-tidy  ../src/**/*.cpp $CHECK_FLAGS $ERR_FLAGS
```
Reference: https://clang.llvm.org/extra/clang-tidy/

### CODE REVIEW

**Comment by hungpham2511** [comment]
Hi @jmirabel, please note that the repo now has run CI jobs on the cpp source code. One for building / testing and one for linting.

I will merge this PR later today if we all agree on the linting / build config.

As a side node, for #86 you will need to amend the CI steps to run pinocchio and qpoases.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    - uses: actions/checkout@v2
+
+    - name: Install dep
+      run: |
+        sudo apt-get install -y clang-tidy libeigen3-dev
+
+    - name: Build
+      run: |
+        cd $GITHUB_WORKSPACE/cpp && mkdir build && cd build
+        cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 ..
```
Comment (jmirabel) [inline .github/workflows/main.yml:24]: Any reason not to call make here ?

File: cpp/src/toppra/constraint.cpp [hunk 2]
```diff
@@ -33,46 +33,64 @@ void allocateLinearPart(std::size_t N, Eigen::Index k, Eigen::Index m,
...
+  for (Vector& x : a) { x.resize(m);
+}
+  for (Vector& x : b) { x.resize(m);
+}
+  for (Vector& x : c) { x.resize(m);
+}
+  for (Matrix& x : F) { x.resize(k, m);
+}
+  for (Vector& x : g) { x.resize(k);
+}
```
Comment (jmirabel) [inline cpp/src/toppra/constraint.cpp:45]: Is it a linter who introduced those changes ?
I find the syntax very misleading. I have nothing against adding brackets. I can understand
```cpp
for (Vector& x : a) {
  x.resize(m);
}
```
```cpp
for (Vector& x : a) { x.resize(m); }
```
but the current result, I cannot understand the motivation for it.

Comment (jmirabel) [reply cpp/src/toppra/constraint.cpp:45]: And more important, I don't understand why the closing bracket is not indented as the for.

File: cpp/src/toppra/constraint.cpp [hunk 3]
```diff
@@ -157,15 +176,17 @@ void LinearConstraint::allocateParams(std::size_t N,
...
+  if (hasXbounds()) {
     xbound.resize(N);
 }
+}
 
 void LinearConstraint::computeParams(const GeometricPath& path, const Vector& gridpoints,
     Vectors& a, Vectors& b, Vectors& c, Matrices& F, Vectors& g,
-    Bounds ubound, Bounds& xbound)
+    const Bounds& ubound, Bounds& xbound)
```
Comment (jmirabel) [inline cpp/src/toppra/constraint.cpp:189]: This should be `Bounds& ubound, Bounds& xbound)`

File: cpp/src/toppra/constraint.hpp [hunk 4]
```diff
@@ -125,7 +125,7 @@ class LinearConstraint {
     void computeParams(const GeometricPath& path, const Vector& gridpoints,
         Vectors& a, Vectors& b, Vectors& c,
         Matrices& F, Vectors& g,
-        Bounds ubound, Bounds& xbound);
+        const Bounds& ubound, Bounds& xbound);
```
Comment (jmirabel) [inline cpp/src/toppra/constraint.hpp:128]:
```suggestion
        Bounds& ubound, Bounds& xbound);
```

File: cpp/src/toppra/constraint/linear_joint_acceleration.cpp [hunk 5]
```diff
@@ -13,15 +13,16 @@ std::ostream& LinearJointAcceleration::print (std::ostream& os) const
...
     throw std::invalid_argument("Bad acceleration limits.");
 }
+}
 
 void LinearJointAcceleration::computeParams_impl(const GeometricPath& path,
         const Vector& gridpoints,
         Vectors& a, Vectors& b, Vectors& c,
         Matrices& F, Vectors& g,
-        Bounds, Bounds&)
+        Bounds /*ubound*/, Bounds& /*xbound*/)
```
Comment (jmirabel) [inline cpp/src/toppra/constraint/linear_joint_acceleration.cpp:25]:
```suggestion
        Bounds& /*ubound*/, Bounds& /*xbound*/)
```

File: cpp/src/toppra/constraint/linear_joint_velocity.cpp [hunk 6]
```diff
@@ -13,16 +13,18 @@ std::ostream& LinearJointVelocity::print (std::ostream& os) const
...
     throw std::invalid_argument("Bad velocity limits.");
 }
+}
 
 void LinearJointVelocity::computeParams_impl(const GeometricPath& path,
         const Vector& gridpoints,
-        Vectors&, Vectors&, Vectors&, Matrices&, Vectors&,
-        Bounds , Bounds& xbound)
+        Vectors& /*a*/, Vectors& /*b*/, Vectors& /*c*/, Matrices& /*F*/, Vectors& /*g*/,
+        Bounds  /*ubound*/, Bounds& xbound)
```
Comment (jmirabel) [inline cpp/src/toppra/constraint/linear_joint_velocity.cpp:27]:
```suggestion
        Bounds& /*ubound*/, Bounds& xbound)
```

File: cpp/src/toppra/geometric_path.hpp [hunk 7]
```diff
@@ -52,7 +52,7 @@ class PiecewisePolyPath : public GeometricPath {
    * @param coefficients Polynoamial coefficients.
    * @param breakpoints Vector of breakpoints.
    */
-  PiecewisePolyPath(const Matrices &, const std::vector<value_type> &);
+  PiecewisePolyPath(const Matrices &, std::vector<value_type> );
```
Comment (jmirabel) [inline cpp/src/toppra/geometric_path.hpp:55]: Why don't you want to pass a const reference ?

File: cpp/src/toppra/constraint.cpp (hunk 3 above)
Comment (jmirabel) [inline cpp/src/toppra/constraint.cpp:189]:
```suggestion
    Bounds& ubound, Bounds& xbound)
```

File: cpp/src/toppra/constraint.cpp (hunk 2 above)
Comment (hungpham2511) [reply cpp/src/toppra/constraint.cpp:45]: This commit is from the linter. Do you suggest to change all to the below? I can make that change.

```
for (Vector& x : a) { x.resize(m); }
```

File: cpp/src/toppra/geometric_path.hpp (hunk 7 above)
Comment (hungpham2511) [reply cpp/src/toppra/geometric_path.hpp:55]: Hm. This looks quite unfamiliar. I should have checked the linter fixes more carefully.

File: .github/workflows/main.yml (hunk 1 above)
Comment (hungpham2511) [reply .github/workflows/main.yml:24]: Build and test are in the circleci CI step. 

https://circleci.com/gh/hungpham2511/toppra/775?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: cpp/src/toppra/geometric_path.hpp (hunk 7 above)
Comment (hungpham2511) [reply cpp/src/toppra/geometric_path.hpp:55]: I wonder why it changes const reference to pass-by-value though.

File: cpp/src/toppra/constraint.cpp (hunk 3 above)
Comment (hungpham2511) [reply cpp/src/toppra/constraint.cpp:189]: Noted, will change.

File: cpp/src/toppra/geometric_path.hpp (hunk 7 above)
Comment (jmirabel) [reply cpp/src/toppra/geometric_path.hpp:55]: I came to the same conclusion. But if they are the same, I don't understand why it changed it (and, even more, why did it decide to use a C++11 only feature when a C++98 do the same job)... But there might be a good reason. Anyway, I let you keep the one you like.

File: cpp/src/toppra/constraint.cpp (hunk 3 above)
Comment (jmirabel) [reply cpp/src/toppra/constraint.cpp:189]: I realized later that those error were probably mine.

File: cpp/src/toppra/constraint.cpp (hunk 2 above)
Comment (jmirabel) [reply cpp/src/toppra/constraint.cpp:45]: You mean you will change the linter rules or you will manually do it ?
There are many of them and, if you change them manually, the linter will complain again. In some project in my lab, we use clang-format, with google-style guide. It does not completely match my taste but I have nothing better to propose (EDIT there are some other proposed here https://clang.llvm.org/docs/ClangFormatStyleOptions.html).

Comment (hungpham2511) [reply cpp/src/toppra/constraint.cpp:45]: I use clang-tidy, which I suppose only works on the logic level. Style-wise it doesn't care. Will eventually add a config for clang-format and enforce it. 

I don't have strong opinion on this matter. As long as the code is nicely formatted and consistent I am ok.

Comment (jmirabel) [reply cpp/src/toppra/constraint.cpp:45]: `for (Vector& x : a) { x.resize(m); }` is fine for me.
Mind that there are plenty of places where clang-tidy introduced a badly indented closing bracket.

I let you choose whether or not to use clang-format. I usually don't see it as necessary. I can see that we did not follow the same convention in many places (not restricted for formatting) anyway, so I guess formatting does not make much sense.

Comment (hungpham2511) [reply cpp/src/toppra/constraint.cpp:45]: I think we should have a consistent format style, and clang-format is the easiest way to do it. 

The only problem with introducing clang-format now is merge conflicts, so I will not format the code you are working on. However, I will run that on code I am working on.

**Comment by hungpham2511** [comment]
Hi @jmirabel Following our discussion, I revert all the changes made the linter. Also, add a clang-format configuration, enforcing is optional for now.

**Comment by jmirabel** [comment]
Thanks.

**Pull request merged**